File: java/testsuite/OfflineExamples/PolygonTriangulation/src/io/PolygonReader.java
package io;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

import vsdk.toolkit.environment.geometry.surface.polygon.Polygon2D;

public class PolygonReader
{
    public Polygon2D read(String fileName) throws IOException
    {
        List<String> lines = Files.readAllLines(Path.of(fileName));
        List<String> tokens = new ArrayList<>();

        for ( String line : lines ) {
            String trimmedLine = line.trim();
            if ( trimmedLine.isEmpty() ) {
                continue;
            }
            for ( String token : trimmedLine.split("\\s+") ) {
                if ( !token.isEmpty() ) {
                    tokens.add(token);
                }
            }
        }

        int tokenIndex = 0;
        int contourCount = Integer.parseInt(tokens.get(tokenIndex++));
        Polygon2D polygon = new Polygon2D();
        polygon.loops.clear();

        for ( int contourIndex = 0; contourIndex < contourCount; contourIndex++ ) {
            polygon.nextLoop();
            int pointCount = Integer.parseInt(tokens.get(tokenIndex++));
            for ( int pointIndex = 0; pointIndex < pointCount; pointIndex++ ) {
                double x = Double.parseDouble(tokens.get(tokenIndex++));
                double y = Double.parseDouble(tokens.get(tokenIndex++));
                polygon.addVertex(x, y);
            }
        }

        if ( !polygon.loops.isEmpty() && polygon.loops.get(0).vertices.isEmpty() ) {
            polygon.loops.remove(0);
        }

        return polygon;
    }
}

File: typescript/testsuite/OfflineExamples/PolygonTriangulation/src/io/PolygonReader.ts
// Basic Node classes
import { readFileSync } from "node:fs";

// VSDK classes
import { Double, Integer, Polygon2D } from "@vitral/base";

export class PolygonReader {
    public read(fileName: string): Polygon2D {
        const lines: string[] = readFileSync(fileName, "utf8").split("\n");
        const tokens: string[] = [];

        for (const line of lines) {
            const trimmedLine: string = line.trim();
            if (trimmedLine.length === 0) {
                continue;
            }
            for (const token of trimmedLine.split(/\s+/u)) {
                if (token.length > 0) {
                    tokens.push(token);
                }
            }
        }

        let tokenIndex = 0;
        const contourCount: number = Integer.parseInt(tokens[tokenIndex++]!);
        const polygon: Polygon2D = new Polygon2D();
        polygon.loops.length = 0;

        let contourIndex: number;
        for (contourIndex = 0; contourIndex < contourCount; contourIndex++) {
            polygon.nextLoop();
            const pointCount: number = Integer.parseInt(tokens[tokenIndex++]!);
            let pointIndex: number;
            for (pointIndex = 0; pointIndex < pointCount; pointIndex++) {
                const x: number = Double.parseDouble(tokens[tokenIndex++]!);
                const y: number = Double.parseDouble(tokens[tokenIndex++]!);
                polygon.addVertex(x, y);
            }
        }

        if (polygon.loops.length > 0 && polygon.loops[0]!.vertices.length === 0) {
            polygon.loops.splice(0, 1);
        }

        return polygon;
    }
}
